### PR TITLE
cmd/utils, p2p: clean up discovery setup

### PR DESCRIPTION
**Pull request by lightclient**
While reading through the server, I noticed the `setupDiscovery()` function had a lot of branching and it seemed like an somewhat easy thing to simplify. It appears that the complication stems mostly from the fact that `--nodiscover` actually doesn't fully stop discovery. There are two cases where it will proceed:

1) ~~DNS discovery will currently proceed regardless of this flag, because
   it uses this new per-protocol discovery mechanism via the discmix
   iterator and it is set before the early return.~~ Turns out this is actually not true, the dns URLs are nil'd out in the flags: https://discord.com/channels/420394352083337236/424272951366385664/1120690530792255548

2) If light sync or server is specified, the flags interpreter will
   forcefully start discv5. This is, I think, why the function needed to
   be more complicated.

After this change, the DNS discovery will not proceed when `--nodiscover` is set. ~~This made me think that it might be interesting to give users more control over their discovery mechanisms, e.g. DNS-only, DHT-only, etc~~ (I think this will be possible now with the `discv4` flag). Also, the geth command will exit if a light-mode flag is used with `--nodiscover`.

That should make it so `--nodiscover` really does mean no discovery!

### CODE REVIEW

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -765,6 +765,12 @@ var (
 		Usage:    "Disables the peer discovery mechanism (manual peer addition)",
 		Category: flags.NetworkingCategory,
 	}
+	DiscoveryV4Flag = &cli.BoolFlag{
+		Name:     "v4disc",
```
Comment (lightclient) [inline cmd/utils/flags.go:769]: I like `discv4` and `discv5` better as flags. Thoughts on marking `v5disc` deprecated and add a new one with the same functionality where the version is a postfix?

Comment (fjl) [reply cmd/utils/flags.go:769]: why not!

Comment (lightclient) [reply cmd/utils/flags.go]: done in 9ec6f5bc11115dcdd2198d05cae1fefc85e6bc65

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -766,13 +766,18 @@ var (
...
-		Name:     "v4disc",
+		Name:     "discv4",
 		Usage:    "Enables the V4 discovery mechanism",
 		Category: flags.NetworkingCategory,
 		Value:    true,
 	}
-	DiscoveryV5Flag = &cli.BoolFlag{
+	LegacyDiscoveryV5Flag = &cli.BoolFlag{
 		Name:     "v5disc",
+		Usage:    "Enables the experimental RLPx V5 (Topic Discovery) mechanism (note: this flag is deprecate, please use --discv5 instead)",
```
Comment (fjl) [inline cmd/utils/flags.go:776]: d

Comment (lightclient) [reply cmd/utils/flags.go]: oops

File: p2p/server.go [hunk 3]
```diff
@@ -538,57 +541,31 @@ func (srv *Server) setupLocalNode() error {
...
-	if err != nil {
-		return err
+
+	var (
+		sconn     discover.UDPConn = conn
+		unhandled chan discover.ReadPacket
+	)
+	// Only initialize unhandled channel if discv5 is running.
+	if srv.DiscoveryV5 {
+		unhandled = make(chan discover.ReadPacket, 100)
```
[the remark this answers is not in the document]
Comment (lightclient) [reply p2p/server.go:559]: Ah good point. Updated.

**Pull request merged**